src/graph.rs

```diff
@@ -502,25 +497,64 @@ fn is_false(v: &bool) -> bool {
   !v
 }
 
-#[derive(Debug, Default, Clone, Serialize, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ImportError {
+  TypeAssertionFailed {
+    specifier: ModuleSpecifier,
+    range: Range,
+    actual_media_type: MediaType,
+    expected_media_type: MediaType,
+  },
+  TypeAssertionUnsupported {
+    specifier: String,
+    range: Range,
+    assert_type: String,
+  },
+}
+
+impl std::error::Error for ImportError {}
+
+impl fmt::Display for ImportError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match &self {
+      ImportError::TypeAssertionFailed { specifier, actual_media_type: MediaType::Json, expected_media_type, .. } => write!(f, "Expected a {expected_media_type} module, but identified a Json module. Consider importing Json modules with an import assertion with the type of \"json\".\n  Specifier: {specifier}"),
+      ImportError::TypeAssertionFailed { specifier, actual_media_type, expected_media_type, .. } => write!(f, "Expected a {expected_media_type} module, but identified a {actual_media_type} module.\n  Specifier: {specifier}"),
+      ImportError::TypeAssertionUnsupported { assert_type, .. } => write!(f, "The import assertion type of \"{assert_type}\" is unsupported."),
+    }
+  }
+}
+
+impl Serialize for ImportError {
+  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+  where
+    S: Serializer,
+  {
+    serializer.serialize_str(&self.to_string())
+  }
+}
+
+#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
 #[serde(rename_all = "camelCase")]
+pub struct Import {
+  pub specifier: String,
+  pub range: Range,
+  #[serde(skip_serializing_if = "is_false")]
+  pub is_dynamic: bool,
+  #[serde(skip_serializing_if = "ImportAssertions::is_empty")]
+  pub assertions: ImportAssertions,
+  // Note: This is only populated in a final pass at the end of graph building.
+  #[serde(skip_serializing_if = "Vec::is_empty")]
+  pub errors: Vec<ImportError>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
 pub struct Dependency {
-  #[serde(
-    rename = "code",
-    skip_serializing_if = "Resolution::is_none",
-    serialize_with = "serialize_resolution"
-  )]
   pub maybe_code: Resolution,
-  #[serde(
-    rename = "type",
-    skip_serializing_if = "Resolution::is_none",
-    serialize_with = "serialize_resolution"
-  )]
   pub maybe_type: Resolution,
-  #[serde(skip_serializing_if = "is_false")]
   pub is_dynamic: bool,
-  #[serde(rename = "assertType", skip_serializing_if = "Option::is_none")]
-  pub maybe_assert_type: Option<String>,
+  /// List of code imports which refer to this dependency. Only ones which can
+  /// populate `Dependency::maybe_code`.
+  pub imports: Vec<Import>,
```

Comment:
I'm not sure if "imports" is the right word here because currently in this PR this could be imports, exports, or typescript references. Perhaps it should be `specifiers` instead?

Reply:
Ok, imports doesn't sound too bad. You're right that it's more clear.